### PR TITLE
Update transpose optimization to be more QDQ aware

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

Rework some aspects of the transpose optimizer to ensure we have valid QDQ node units when it is done.

Conceptually we need to let individual Transpose nodes move through the graph when optimizing. That can invalidate existing QDQ node units or require new ones. We can fix this after inserting new nodes, or when transpose optimization finishes moving Transpose nodes.

Fix when inserting new node
- TransposeInputs can add an Unsqueeze (to broadcast) and Transpose to a node's inputs
  - if there was a DQ node providing the input, add a Q -> DQ after inserting the Unsqueeze/Transpose to make a QDQ node unit for the new node. 
    - Unsqueeze/Transpose don't change data, so we can copy the type/scale/zero point from the existing DQ

Fixes when transpose optimization completes moving Transpose nodes
- Remove empty DQ -> Q pairs if the type/scale/zero point match
  - Pushing a Transpose through may have resulted in an existing Transpose/Reshape being cancelled and removed leaving an empty QDQ node unit
  - the Transpose being moved may have started in a QDQ node unit
- Transpose that got blocked inside existing QDQ node unit
  - e.g. if we hit a DQ -> MatMul -> Q node unit the Transpose gets blocked after the DQ
  - insert a Q -> DQ after the Transpose to put it in a QDQ node unit and repair the original QDQ node unit
- Transpose moves past a DQ providing a graph output
  - insert a Q -> DQ so the Transpose is in a QDQ node unit

This replaces the existing phase 2 logic which flipped a DQ -> Transpose to fix a broken QDQ node unit. The new approach should handle more scenarios and hopefully produce a better graph. 

Additionally the logic to handle updates to shared initializers that feed DQ nodes was simplified (i.e. largely removed). When we update the shared initializer a Squeeze (if broadcast) and Transpose is added between the initializer and the DQ for other usages of it. We only need to check for this pattern in EstimateTransposeValueCost by looking past a DQ node. We do not need to track the individual DQ nodes leading to an updated shared initializer.

### Motivation and Context

Initially to fix QNN issue with non-const input being transpose and the QDQ node units being broken.

### CODE REVIEW

File: onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc [hunk 1]
```diff
@@ -95,32 +98,104 @@ static std::unique_ptr<api::NodeRef> MakeSqueezeOrUnsqueeze(int64_t opset, api::
   return graph.AddNode(op_type, inputs, /*num_outputs*/ 1);
 }
 
+// Creates a QuantizeLinear or Quantize node. Does not update output ValueInfo.
```
Comment (askhade) [inline onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:101]:
```suggestion
// Creates a QuantizeLinear or Dequantize node. Does not update output ValueInfo.
```

**Comment by askhade** [comment]
LGTM

**Review by askhade — APPROVED** [review]
LGTM!!

**Comment by adrianlizarraga** [comment]
@skottmckay The onnxruntime-tvm checks failed for this PR: https://github.com/microsoft/onnxruntime/actions/runs/6943126715/job/18887531996

image

**Pull request merged**